packages/@aws-cdk/aws-batch/lib/compute-environment.ts

```diff
@@ -0,0 +1,484 @@
+      throw new Error('computeResources is missing but required on an unmanaged compute environment');
+    }
+
+    // Setting a bid percentage is only allowed on SPOT resources +
+    // Cannot use SPOT_CAPACITY_OPTIMIZED when using ON_DEMAND
+    if (props.computeResources) {
+      if (props.computeResources.type === ComputeResourceType.ON_DEMAND) {
+        // VALIDATE FOR ON_DEMAND
+
+        // Bid percentage is not allowed
+        if (props.computeResources.bidPercentage !== undefined) {
+          throw new Error('Setting the bid percentage is only allowed for SPOT type resources on a batch compute environment');
+        }
+
+        // SPOT_CAPACITY_OPTIMIZED allocation is not allowed
+        if (props.allocationStrategy && props.allocationStrategy === AllocationStrategy.SPOT_CAPACITY_OPTIMIZED) {
+          throw new Error('The SPOT_CAPACITY_OPTIMIZED allocation strategy is only allowed if the environment is a SPOT type compute environment');
+        }
+      } else {
+        // VALIDATE FOR SPOT
+
+        // Bid percentage must be from 0 - 100
+        if (props.computeResources.bidPercentage !== undefined &&
+          (props.computeResources.bidPercentage < 0 || props.computeResources.bidPercentage > 100)) {
+            throw new Error('Bid percentage can only be a value between 0 and 100');
+        }
+      }
+
+      if (props.computeResources.minvCpus) {
+        // minvCpus cannot be less than 0
+        if (props.computeResources.minvCpus < 0) {
+          throw new Error('Minimum vCpus for a batch compute environment cannot be less than 0');
+        }
+
+        // minvCpus cannot exceed max vCpus
+        if (props.computeResources.maxvCpus &&
+          props.computeResources.minvCpus > props.computeResources.maxvCpus) {
+            throw new Error('Minimum vCpus cannot be greater than the maximum vCpus');
+        }
+      }
+    }
+  }
+
+  private buildInstanceTypes(instanceTypes?: ec2.InstanceType[]): string[] {
+    if (instanceTypes === undefined) {
+      return [
+        'optimal',
+      ];
+    }
+
+    return instanceTypes.map((type: ec2.InstanceType) => type.toString());
+  }
+
+  private buildSecurityGroupIds(vpc: ec2.IVpc, securityGroups?: ec2.ISecurityGroup[]): string[] | undefined {
+    if (securityGroups === undefined) {
+      return [
+        new ec2.SecurityGroup(this, 'Resource-Security-Group', { vpc }).securityGroupId,
+      ];
+    }
+
+    return securityGroups.map((group: ec2.ISecurityGroup) => group.securityGroupId);
+  }
+
+  /**
+   * Generates an AWS IAM role for provisioning spotfleet resources
+   * if the allocation strategy is set to BEST_FIT or not defined.
+   *
+   * @param props - the compute environment construct properties
+   */
+  private getSpotFleetRole(props: ComputeEnvironmentProps): iam.IRole | undefined {
+    const spotFleetArn = `arn:aws:iam::${this.stack.account}:role/aws-service-role/spotfleet.amazonaws.com/AWSServiceRoleForEC2SpotFleet`;
```

Comment:
Is this some magic role that's always present in your account? (Also, I think you want `${Aws.PARTITION}` instead of just `:aws:` there)

Reply:
If you scroll to the bottom here >> https://docs.aws.amazon.com/zh_en/batch/latest/userguide/spot_fleet_IAM_role.html

It is always a role that has to exist before you can make a batch compute environment. I gave it a try to make my own service linked role with similar permissions buts its not as "easy". Do you know if its just possible to make any IAM role with similar permissions and just assign it?

Reply:
So, that doc talks about 2 service-linked roles that are needed:

* `AWSServiceRoleForEC2Spot`
* `AWSServiceRoleForEC2SpotFleet`

The problem is that service-linked roles are unique in an account (because they use the names), so we can't really create them automatically like we do regular roles. We can have an optional parameter that tries to create these... but you don't have to do this in this PR.